// @emergentbase/visual-edits/dist/vite-plugin.mjs
// src/vite-plugin.ts
import path from "path";
import fs from "fs";
import { fileURLToPath } from "url";
import { createRequire } from "module";
var _dirname = path.dirname(fileURLToPath(import.meta.url));
var _require = createRequire(import.meta.url);
function visualEdits(options) {
  let babelTransformSync;
  let babelPlugin;
  return {
    name: "visual-edits",
    enforce: "pre",
    apply(_config, { command }) {
      if (options?.enable === false) return false;
      if (options?.enable === true) return true;
      return command === "serve";
    },
    // 1. Babel metadata plugin — stamp x-* attributes on JSX elements
    transform(code, id) {
      const cleanId = id.replace(/\?.*$/, "");
      if (!/\.[jt]sx$/.test(cleanId)) return null;
      if (cleanId.includes("node_modules") || id.includes("\0")) return null;
      if (!babelTransformSync) {
        babelTransformSync = _require("@babel/core").transformSync;
        babelPlugin = _require("@emergentbase/visual-edits/babel-plugin").default;
      }
      const isTypeScript = cleanId.endsWith(".tsx") || cleanId.endsWith(".ts");
      const result = babelTransformSync(code, {
        filename: cleanId,
        plugins: [babelPlugin],
        parserOpts: {
          plugins: [
            "jsx",
            ...isTypeScript ? ["typescript"] : [],
            "decorators-legacy",
            "classProperties"
          ]
        },
        sourceMaps: true,
        configFile: false,
        babelrc: false,
        presets: []
      });
      return result ? { code: result.code, map: result.map } : null;
    },
    // 2. Inject iframe-guard script into HTML
    transformIndexHtml(html) {
      const tailwindCdn = options?.tailwindCdn ?? "https://cdn.tailwindcss.com";
      const tailwindSnippet = tailwindCdn ? `window.tailwind=window.tailwind||{};tailwind.config={corePlugins:{preflight:false}};var t=document.createElement("script");t.src="${tailwindCdn}";document.head.appendChild(t);` : "";
      const snippet = `<script>
if(window.self!==window.top){
  var s=document.createElement("script");s.src="/visual-edit-overlay.js";document.head.appendChild(s);
  ${tailwindSnippet}
}
</script>`;
      return html.replace("</head>", snippet + "\n</head>");
    },
    // 3. Dev server middleware — overlay serving + shared route middleware
    configureServer(server) {
      const overlayPath = path.resolve(_dirname, "../dist/visual-edit-overlay.js");
      let overlaySource;
      try {
        overlaySource = fs.readFileSync(overlayPath, "utf-8");
      } catch (err) {
        console.error(
          `[visual-edits] Failed to read overlay: ${err.message}`
        );
        overlaySource = "// visual-edit-overlay.js not found";
      }
      server.middlewares.use("/visual-edit-overlay.js", (_req, res) => {
        res.setHeader("Content-Type", "application/javascript");
        res.setHeader("Cache-Control", "no-cache");
        res.end(overlaySource);
      });
      const { createMiddleware } = _require("@emergentbase/visual-edits/server");
      server.middlewares.use(createMiddleware());
    }
  };
}
var vite_plugin_default = visualEdits;
export {
  vite_plugin_default as default,
  visualEdits
};
